### PR TITLE
Infer `uaddr` pointer type from ELF symbol size

### LINKED ISSUE #939: Infer type of uaddr pointer
Take the code below for example:

```d
uprobe:node:_ZN2v88internal17Builtin_JsonParseEiPPNS0_6ObjectEPNS0_7IsolateE
{
  $stringLengthOffset = *uaddr("v8dbg_class_String__length__SMI");
  printf("v8dbg_class_String__length__SMI: %lx\n", $stringLengthOffset);
}
```

`v8dbg_class_String__length__SMI` is an `int` global variable which will always be set to 16 (0x10). But if we run a Node.js program traced by the program above, we'll get the following output:

```console
$ bpftrace test.bt
v8dbg_class_String__length__SMI: 1800000010
```

This happens because the return of `uaddr("v8dbg_class_String__length__SMI")` is an `int64_t *`, but the type should be `int32_t *`. To workaround this I had to cast it to `int32` after dereferencing the pointer, which is not all that intuitive.

There are two things we can do to improve this (and we probably should do both):

1. Infer the pointer type of uaddr if the data size matches a known integer type
2. Allow users to define the pointer type when calling uaddr (`uaddr(name[, type])`). As a bonus, we can use semantic analysis to force users to define a pointer type when the data size doesn't match a known integer type

### PR DESCRIPTION
**Pull request by fbs**
Infer `uaddr` pointer type from ELF symbol size

The size of `uaddr` was hardcoded to 8 bytes. This can lead to
accidentally reading too much data from types smaller than 8 bytes. E.g.
with the following C globals:

```
unsigned int GLOBALA = 0x55555555;
unsigned int GLOBALB = 0x88888888;
unsigned int GLOBALC = -1;
```

Dereferencing `GLOBALA` will also read the content of `GLOBALB`, which
is unexpected and can lead to hard to miss errors::

```
$ bpftrace -e 'u:./a.out:main { printf("%lx\n", *uaddr("GLOBALA")); }'
8888888855555555
```

In this case the pointer would have to be casted to get the expected
behaviour (`*(uint32*)uaddr("GLOBAL")`).

This patch makes use of the ELF object size to determine the symbol
size. If the object size is a known integer size (1,2,4,8 bytes) that is
used, in all other cases 8 is used as size (sticking to the original
behaviour).

If multiple attach points are used the symbol for each of the attach
points must resolve to the same size.

Fixes #939

### CODE REVIEW

**Comment by fbs** [comment]
Looks like disabling PIE (to avoid aslr) for the testsuite doesn't work yet. I'll look into it

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -167,10 +170,14 @@ file(GLOB testprogs testprogs/*.c)
...
   add_executable (${bin_name} ${testprog})
+
+  set_property(TARGET ${bin_name} PROPERTY POSITION_INDEPENDENT_CODE FALSE)
+  target_compile_options(${bin_name} PRIVATE -no-pie)
+
   if(HAVE_SYSTEMTAP_SYS_SDT_H)
     target_compile_definitions(${bin_name} PRIVATE HAVE_SYSTEMTAP_SYS_SDT_H)
   endif(HAVE_SYSTEMTAP_SYS_SDT_H)
-  set_target_properties( ${bin_name} PROPERTIES RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/testprogs/ COMPILE_FLAGS "-g -O0")
+  set_target_properties( ${bin_name} PROPERTIES RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/testprogs/ COMPILE_FLAGS "-g -O0 -no-pie")
```
Comment (mmisono) [inline tests/CMakeLists.txt]: I don't know much about CMake, but I think -no-pie link flag is needed.

```suggestion
  set_target_properties( ${bin_name} PROPERTIES RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/testprogs/ COMPILE_FLAGS "-g -O0" LINK_FLAGS "-no-pie")
```

(It seems  `POSITION_INDEPENDENT_CODE` affects link flags only in [CMake 3.14 and newer](https://cmake.org/cmake/help/latest/policy/CMP0083.html))

File: tests/runtime/call [hunk 2]
```diff
@@ -108,3 +108,8 @@ NAME cat
 RUN bpftrace -v -e 'i:ms:1 { cat("/proc/uptime"); exit();}'
 EXPECT [0-9]*.[0-9]* [0-9]*.[0-9]*
 TIMEOUT 5
+
+NAME uaddr
+RUN bpftrace -v -e 'uprobe:./testprogs/uprobe_test:function1 { printf("%lx\n", *uaddr("GLOBAL_A")); exit(); }' -c ./testprogs/uprobe_test
```
Comment (mmisono) [inline tests/runtime/call]:
```suggestion
RUN bpftrace -v -e 'uprobe:./testprogs/uprobe_test:function1 { printf("0x%lx\n", *uaddr("GLOBAL_A")); exit(); }' -c ./testprogs/uprobe_test
```

File: tests/runtime/call [hunk 3]
```diff
@@ -108,3 +108,8 @@ NAME cat
...
+EXPECT 0x555555555
```
Comment (mmisono) [inline tests/runtime/call]:
```suggestion
EXPECT 0x55555555
```

File: src/ast/semantic_analyser.cpp [hunk 4]
```diff
@@ -478,14 +478,54 @@ void SemanticAnalyser::visit(Call &call)
...
+      struct symbol sym = {};
+      int err = bpftrace_.resolve_uname(name, &sym, ap->target);
+      if (err < 0 || sym.address == 0) {
+        bpftrace_.error(err_, call.loc,
+                        "Could not resolve symbol: " + ap->target + ":" + name);
+      }
+      sizes.push_back(sym.size);
+    }
+
+    for (size_t i = 0; i < sizes.size(); i++) {
```
Comment (mmisono) [inline src/ast/semantic_analyser.cpp]: just a minor comment...
```suggestion
    for (size_t i = 1; i < sizes.size(); i++) {
```

**Comment by fbs** [comment]
thanks @mmisono

File: tests/mocks.h [hunk 5]
```diff
@@ -12,6 +13,8 @@ class MockBPFtrace : public BPFtrace {
       std::unique_ptr<std::istream>(int pid, const std::string &target));
   MOCK_CONST_METHOD1(extract_func_symbols_from_path,
       std::string(const std::string &path));
+  /* MOCK_CONST_METHOD3(resolve_uname, int(const std::string &name, struct symbol * sym, const std::string &path)); */
```
Comment (danobi) [inline tests/mocks.h]: Is the comment necessary?

File: tests/mocks.h [hunk 6]
```diff
@@ -20,6 +23,22 @@ class MockBPFtrace : public BPFtrace {
   {
     return special_probes_;
   }
+
+  int resolve_uname(const std::string& name, struct symbol *sym, const std::string &path) const
```
Comment (danobi) [inline tests/mocks.h]: Not super important but defining this in mocks.cpp is probably better.

Also:
```suggestion
  int resolve_uname(const std::string& name, struct symbol *sym, const std::string &path) const override
```

File: tests/runtime/call [hunk 7]
```diff
@@ -108,3 +108,8 @@ NAME cat
 RUN bpftrace -v -e 'i:ms:1 { cat("/proc/uptime"); exit();}'
 EXPECT [0-9]*.[0-9]* [0-9]*.[0-9]*
 TIMEOUT 5
+
+NAME uaddr
+RUN bpftrace -v -e 'uprobe:./testprogs/uprobe_test:function1 { printf("0x%lx\n", *uaddr("GLOBAL_A")); exit(); }' -c ./testprogs/uprobe_test
```
Comment (danobi) [inline tests/runtime/call]: Might as well also test GLOBAL_[BCD], right?

File: docs/reference_guide.md [hunk 8]
```diff
@@ -1709,7 +1709,20 @@ const char *usbcore_name = "usbcore";
 
 ## 9. `uaddr()`: Address resolution, user-level
 
-Syntax: `uaddr(char *name)`
+Syntax: `u64* uaddr(symbol)`
```
Comment (ajor) [inline docs/reference_guide.md]: It doesn't always return a `u64*` now with these changes, right?

Comment (fbs) [reply docs/reference_guide.md]: Nope it depends. Any suggestions for a better way to write it down? I think that having the return type in the `syntax` is useful but when we do it it must be correct

File: tests/mocks.h (hunk 5 above)
Comment (fbs) [reply tests/mocks.h]: woops, no

File: docs/reference_guide.md (hunk 8 above)
Comment (fbs) [reply docs/reference_guide.md]: @ajor poke

Comment (ajor) [reply docs/reference_guide.md]: Maybe we could list all the forms? e.g.
```
u64 *uaddr(symbol)
u32 *uaddr(symbol)
u16 *uaddr(symbol)
u8 *uaddr(symbol)
```

Comment (fbs) [reply docs/reference_guide.md]: Guess that works, what do you think of the new version @ajor

**Comment by fbs** [comment]
I'll merge this in a few days if there are no objections

File: tests/semantic_analyser.cpp [hunk 9]
```diff
@@ -528,12 +528,41 @@ TEST(semantic_analyser, call_kaddr)
 
 TEST(semantic_analyser, call_uaddr)
 {
-  test("kprobe:f { uaddr(\"counter\"); }", 0);
-  test("kprobe:f { uaddr(\"github.com/golang/glog.severityName\"); }", 0);
```
Comment (ajor) [inline tests/semantic_analyser.cpp]: This test was to make sure that Go symbols with non-alpha-numeric characters are allowed by `check_symbol()`. Now that we're actually resolving symbols in the semantic analyser can `check_symbol` just be removed entirely?

Comment (ajor) [reply tests/semantic_analyser.cpp]: Actually maybe we should keep this as a regression test to make sure no one breaks Go symbols in the future, even if we remove `check_symbol`. What do you think?

Comment (fbs) [reply tests/semantic_analyser.cpp]: I'll add them back in.

File: docs/reference_guide.md [hunk 10]
```diff
@@ -1785,7 +1785,26 @@ const char *usbcore_name = "usbcore";
...
+- `u64 *uaddr(symbol)`
+- `u32 *uaddr(symbol)`
+- `u16 *uaddr(symbol)`
+- `u8  *uaddr(symbol)`
+
+Supported Probe Types:
+- u(ret)probes
+- USDT
+
+**Does not work with ASLR, see issue #75**
```
Comment (mmarchini) [inline docs/reference_guide.md]: Can we add a link?

```suggestion
**Does not work with ASLR, see issue [#75](https://github.com/iovisor/bpftrace/75)**
```

File: src/bpftrace.cpp [hunk 11]
```diff
@@ -1592,38 +1592,44 @@ uint64_t BPFtrace::resolve_cgroupid(const std::string &path) const
...
+  std::string call_str =
+      std::string("objdump -tT ") + path + " | grep -w " + sym->name;
   const char *call = call_str.c_str();
   auto result = exec_system(call);
-  return read_address_from_output(result);
+  sym->address = read_address_from_output(result);
+  /* Trying to grab the size from objdump output is not that easy. foreaech_sym
+     has been around for a while, users should switch to that.
+  */
+  sym->size = 8;
```
Comment (mmarchini) [inline src/bpftrace.cpp]: I wonder if a warning would be appropriate.

(or, at some point, we should just drop support for older BCC versions)

File: tests/semantic_analyser.cpp [hunk 12]
```diff
@@ -528,12 +528,42 @@ TEST(semantic_analyser, call_kaddr)
...
-  test("kprobe:f { uaddr(123); }", 1);
-  test("kprobe:f { uaddr(\"?\"); }", 1);
+  test("u:/bin/bash:main { uaddr(\"github.com/golang/glog.severityName\"); }", 0);
+  test("uprobe:/bin/bash:main { uaddr(\"glob_asciirange\"); }", 0);
+  test("u:/bin/bash:main,u:/bin/bash:readline { uaddr(\"glob_asciirange\"); }",
+       0);
+  test("uprobe:/bin/bash:main { @x = uaddr(\"glob_asciirange\"); }", 0);
+  test("uprobe:/bin/bash:main { uaddr(); }", 1);
+  test("uprobe:/bin/bash:main { uaddr(123); }", 1);
+  test("uprobe:/bin/bash:main { uaddr(\"?\"); }", 1);
```
Comment (mmarchini) [inline tests/semantic_analyser.cpp]: Can we also test against var and maps as arguments?

```
  test("uprobe:/bin/bash:main { $str = \"glob_asciirange\"; uaddr($str); }", 1);
  test("uprobe:/bin/bash:main { @str = \"glob_asciirange\"; uaddr(@str); }", 1);
```

File: src/bpftrace.cpp (hunk 11 above)
Comment (fbs) [reply src/bpftrace.cpp]: I'm going to put up a removal PR as the uprobe_offset patches already seem to depend on having the bcc helpers.

**Pull request merged**